### PR TITLE
implement rollout lib

### LINKED ISSUE #295: Add fw rollout module

### PR DESCRIPTION
**Pull request by mroz22**
resolves #295

### CODE REVIEW

File: src/js/core/methods/FirmwareUpdate.js [hunk 1]
```diff
@@ -21,18 +21,7 @@ export default class FirmwareUpdate extends AbstractMethod {
...
-        validateParams(payload, [
-            { name: 'payload', type: 'buffer', obligatory: true },
-            // { name: 'hash', type: 'string' },
-        ]);
-
-        this.params = {
-            payload: payload.payload,
-            length: payload.payload.byteLength,
-        };
+        this.params = {};
```
Comment (szymonlesisz) [inline src/js/core/methods/FirmwareUpdate.js]: add possibility to upload custom Fw without rollout lib

File: src/ts/types/trezor/device.d.ts [hunk 2]
```diff
@@ -73,6 +63,7 @@ export interface Features {
     session_id?: string;
     passphrase_always_on_device?: boolean;
     capabilities?: string[];
+    isBitcoinOnly: boolean;
```
Comment (szymonlesisz) [inline src/ts/types/trezor/device.d.ts]: this shouldnt be here, should be in KnownDevice type

File: src/js/device/Device.js [hunk 3]
```diff
@@ -388,15 +389,17 @@ export default class Device extends EventEmitter {
...
+            // maybe not in bootloade?
+            this.firmwareRelease = getRelease(feat);
         }
         // GetFeatures doesn't return 'session_id'
         if (this.features && this.features.session_id && !feat.session_id) {
             feat.session_id = this.features.session_id;
         }
         this.features = feat;
         this.featuresNeedsReload = false;
+        this.isBitcoinOnly = !!(feat.capabilities && feat.capabilities.length > 0 && !feat.capabilities.includes('Capability_Bitcoin_like'));
```
Comment (szymonlesisz) [inline src/js/device/Device.js]: unnecessary `feat.capabilities.length > 0` check

File: src/js/device/Device.js [hunk 4]
```diff
@@ -388,15 +389,17 @@ export default class Device extends EventEmitter {
         // check if FW version did change
         if (versionCompare(version, this.getVersion()) !== 0) {
             this.unavailableCapabilities = getUnavailableCapabilities(feat, getAllNetworks(), DataManager.getConfig().supportedFirmware);
-            this.firmwareStatus = checkFirmware(version, feat);
-            this.firmwareRelease = getLatestRelease(version);
+            this.firmwareStatus = getFirmwareStatus(feat);
+            // maybe not in bootloade?
```
Comment (szymonlesisz) [inline src/js/device/Device.js]: ?

File: src/js/data/FirmwareInfo.js [hunk 5]
```diff
@@ -1,46 +1,48 @@
...
-            return 'required';
-        } else {
-            return 'outdated';
-        }
+    const info = getInfo({features, releases: releases[features.major_version]});
+
+    // should not happen, possibly if releases list contains inconsistent data or so
+    if (!info) return 'unknown';
+
+    if (info.isRequired) {
```
Comment (szymonlesisz) [inline src/js/data/FirmwareInfo.js]: nitpick: use same syntax as above (!info) `if (info.isRequired) return 'required';`

File: src/js/core/methods/FirmwareUpdate.js [hunk 6]
```diff
@@ -57,14 +62,33 @@ export default class FirmwareUpdate extends AbstractMethod {
...
+            binary = this.params.binary;
+        } else {
+            const firmware = await getBinary({
+                // features and releases are used for sanity checking inside @trezor/rollout
+                features: device.features,
+                releases: DataManager.assets[`firmware-t${device.features.major_version}`],
+                // version argument is used to find and fetch concrete release from releases list
+                version: this.params.version,
+                btcOnly: this.params.btcOnly,
+                baseUrl: 'https://wallet.trezor.io',
```
Comment (szymonlesisz) [inline src/js/core/methods/FirmwareUpdate.js]: not sure if this url should be hardcoded, maybe it should be in config.xml?

File: src/js/device/Device.js (hunk 3 above)
Comment (mroz22) [reply src/js/device/Device.js]: removed, will be implemented consumer-side

**Pull request merged**